core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,275 @@
+   * @param children the children of the {@code path}
+   * @throws IllegalArgumentException when {@code path} already exists or if any child already
+   *                                  exists
+   */
+  public void addChildren(AlluxioURI path, Collection<UfsStatus> children) {
+    UfsStatus status = mStatuses.get(path);
+    // If this path doesn't yet exist, we can't keep track of the parent-child relationship
+    // We can still add statuses to the cache regardless
+    if (status != null) {
+      mChildren.computeIfAbsent(path, ufsStatus -> new ConcurrentHashSet<>()).addAll(children);
+    }
+    children.forEach(child -> {
+      AlluxioURI childPath = path.joinUnsafe(child.getName());
+      addStatus(childPath, child);
+    });
+  }
+
+  /**
+   * Remove a status from the cache.
+   *
+   *  Any children added to this status will remain in the cache.
+   *
+   * @param path the path corresponding to the {@link UfsStatus} to remove
+   * @return the removed UfsStatus
+   */
+  public UfsStatus remove(AlluxioURI path) {
+    UfsStatus removed = mStatuses.remove(path);
+    if (removed == null) {
+      return null;
+    }
+
+    mChildren.remove(path); // ok if there aren't any children
+    return removed;
+  }
+
+  /**
+   * Get the UfsStatus from a given AlluxioURI.
+   *
+   * @param path the path the retrieve
+   * @return The corresponding {@link UfsStatus} or {@code null} if there is none stored
+   */
+  public UfsStatus getStatus(AlluxioURI path) {
+    return mStatuses.get(path);
+  }
+
+  /**
+   * Fetches children of a given alluxio path, stores them in the cache, then returns them.
+   *
+   * Children can be returned in a few ways
+   * 1. Children already exist in the internal index. We simply return them
+   * 2. If children did not already exist in the index, then check if there was a scheduled
+   * prefetch job running for this path. If so, wait for the job to finish and return the result.
+   * 3. If no prefetch job, and children don't yet exist in the cache, then if the fallback
+   * parameter is true, fetch them from the UFS and store them in the cache. Otherwise, simply
+   * return null.
+   *
+   * @param path the Alluxio path to get the children of
+   * @param mountTable the Alluxio mount table
+   * @param useFallback whether or not to fall back to calling the UFS
+   * @return child UFS statuses of the alluxio path, or null if no prefetch job and fallback
+   *         specified as false
+   * @throws InvalidPathException if the alluxio path can't be resolved to a UFS mount
+   */
+  @Nullable
+  public Collection<UfsStatus> fetchChildrenIfAbsent(AlluxioURI path, MountTable mountTable,
+      boolean useFallback)
+      throws IOException, InvalidPathException {
+    Collection<UfsStatus> children = getChildren(path);
+    if (children != null) {
+      return children;
+    }
+    Future<Collection<UfsStatus>> prefetchJob = mActivePrefetchJobs.remove(path);
```

Comment:
Removing it means other callers which call this same method would think there are no prefetch jobs? Is that intended?

Reply:
In the current code path, we never have concurrent callers on the same path, but I will change the code so that it does work for concurrent callers.